code/modules/research/rdmachines.dm

```diff
@@ -17,50 +17,16 @@ var/list/default_material_composition = list(MATERIAL_STEEL = 0, MATERIAL_GLASS
 /obj/machinery/r_n_d/attack_hand(mob/user as mob)
 	return
 
-/obj/machinery/r_n_d/proc/getMaterialType(var/name)
-	switch(name)
-		if(MATERIAL_STEEL)
-			return /obj/item/stack/material/steel
-		if(MATERIAL_GLASS)
-			return /obj/item/stack/material/glass
-		if(MATERIAL_GOLD)
-			return /obj/item/stack/material/gold
-		if(MATERIAL_SILVER)
-			return /obj/item/stack/material/silver
-		if(MATERIAL_PLASMA)
-			return /obj/item/stack/material/plasma
-		if(MATERIAL_URANIUM)
-			return /obj/item/stack/material/uranium
-		if(MATERIAL_DIAMOND)
-			return /obj/item/stack/material/diamond
-	return null
-
-/obj/machinery/r_n_d/proc/getMaterialName(var/type)
-	switch(type)
-		if(/obj/item/stack/material/steel)
-			return MATERIAL_STEEL
-		if(/obj/item/stack/material/glass)
-			return MATERIAL_GLASS
-		if(/obj/item/stack/material/gold)
-			return MATERIAL_GOLD
-		if(/obj/item/stack/material/silver)
-			return MATERIAL_SILVER
-		if(/obj/item/stack/material/plasma)
-			return MATERIAL_PLASMA
-		if(/obj/item/stack/material/uranium)
-			return MATERIAL_URANIUM
-		if(/obj/item/stack/material/diamond)
-			return MATERIAL_DIAMOND
-
 /obj/machinery/r_n_d/proc/eject(var/material, var/amount)
 	if(!(material in materials))
 		return
-	var/obj/item/stack/material/sheetType = getMaterialType(material)
-	var/perUnit = initial(sheetType.perunit)
-	var/eject = round(materials[material] / perUnit)
+	var/eject = materials[material]
```

Comment:
`eject = materials[material]` should be rounded.

Reply:
This list shouldn't contain non integer values

Reply:
```
/obj/machinery/r_n_d/protolathe/RefreshParts()
<...>
	T = 0
	for(var/obj/item/weapon/stock_parts/manipulator/M in component_parts)
		T += M.rating
	mat_efficiency = 1 - (T - 2) / 8
```
Protolathe req 2 manipulators.
For example, get 2 manuputators with rating = 2.
mat_efficiency = 1 - (4 - 2)/8 = 0.75

```
/obj/machinery/r_n_d/protolathe/proc/build(var/datum/design/D)
	<...>
	for(var/M in D.materials)
		materials[M] = max(0, materials[M] - D.materials[M] * mat_efficiency)
```
`D.materials[M]` - integer
`D.materials[M] * 0.75` - can be not integer
`materials[M] - D.materials[M] * mat_efficiency` - can be not integer 

`materials[M]`  - can be not integer